flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -147,6 +147,67 @@ mlir::Value calculateTripCount(fir::FirOpBuilder &builder, mlir::Location loc,
 
   return tripCount;
 }
+
+/// Check if cloning the bounds introduced any dependency on the outer region.
+/// If so, then either clone them as well if they are MemoryEffectFree, or else
+/// copy them to a new temporary and add them to the map and block_argument
+/// lists and replace their uses with the new temporary.
+///
+/// TODO: similar to the above functions, this is copied from OpenMP lowering
+/// (in this case, from `genBodyOfTargetOp`). Once we move to a common lib for
+/// these utils this will move as well.
+void cloneOrMapRegionOutsiders(fir::FirOpBuilder &builder,
+                               mlir::omp::TargetOp targetOp) {
+  mlir::Region &region = targetOp.getRegion();
+  mlir::Block *regionBlock = &region.getBlocks().front();
+  llvm::SetVector<mlir::Value> valuesDefinedAbove;
+  mlir::getUsedValuesDefinedAbove(region, valuesDefinedAbove);
+
+  while (!valuesDefinedAbove.empty()) {
+    for (mlir::Value val : valuesDefinedAbove) {
+      mlir::Operation *valOp = val.getDefiningOp();
+      if (mlir::isMemoryEffectFree(valOp)) {
```

Comment:
The case where the value defined above is a block argument (`valOp == nullptr`) will cause the compiler to crash during this check. If there is a reason why that's never supposed to happen, I'd suggest adding a specific `assert` for it. Otherwise, I think that case should be handled.

Reply:
This is copied from `OpenMP.cpp` and will extracted to a shared util later (see `TODO` above the function). So I assume we did not hit problems with this yet. But I agree with you, I added an `assert` here and `OpenMP.cpp` to be on the safe side.